### PR TITLE
PMP: add a new way to compute vertex normals

### LINKED ISSUE #2047: Bug in PMP::compute_vertex_normal()
## Issue Details

There is a bug in [`PMP::compute_vertex_normal()`][doc]

  [doc]: http://doc.cgal.org/latest/Polygon_mesh_processing/group__PMP__normal__grp.html#ga240ed0a1f5efbea472f998fada07e112

For a vertex of a surface mesh, it computes the normalized normals of all incident faces, and compute the unweighted mean of those normalized normals. ... That is wrong!

Steps to easily see the bug:
- Load the OFF file `demo/Polyhedron/data/cube.off` (that is a simple triangulated unit cube: 8 vertices, 12 faces), in the polyhedron demo,
- Use "create a point set from vertices". That will display the normals.
- For each vertex that is incident to four faces, you will see that the normal is wrong : all vertex normals of the 8 vertices of the cube should be proportional to the diagonal vectors (±1, ±1, ±1).

There are several methods to compute vertex normals of a mesh, and the unweighted mean is clearly always wrong (but if the mesh is very isotropic). See details our three different methods [in this blog][blog]. The page points to articles, one comparing six different methods to compute normals. The best one seems to weight the mean with the angles of the face around the vertex.

  [blog]: http://meshlabstuff.blogspot.fr/2009/04/on-computation-of-vertex-normals.html

## Environment
* CGAL version: current `master` branch (pre-CGAL-4.10-beta1, commit 82286eb79181a54158fdddc4d78e06d56f0e1f1f)

### PR DESCRIPTION
**Pull request by MaelRL**
## Summary of Changes

This PR adds two new ways to compute normals at vertices. 

Instead of defining the normal as the sum of incident face normals, it is computed as the normal that minimizes the maximum distance to the face normals. The default method is changed to be a sine-weighted sum of face normals, and it is used in back-up if we fail to find that normal.

## Release Management

* Affected package(s): `Polygon_mesh_processing`
* Issue(s) solved (if any): fix #2047

### CODE REVIEW

**Comment by lrineau** [comment]
Travis found a lot of compilation errors, in headers:
https://travis-ci.org/CGAL/cgal/jobs/555792898
```
.../Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/compute_normal.h:584:12: error: no member named 'container' in namespace 'boost'
    boost::container::flat_map<face_descriptor, FT> alphas;
    ~~~~~~~^
```
and:
```
In file included from .../Installation/../Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/clip.h:28:
In file included from .../Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/corefinement.h:30:
In file included from .../Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/internal/Corefinement/Visitor.h:30:
In file included from .../Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/internal/Corefinement/face_graph_utils.h:28:
In file included from .../Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orientation.h:31:
.../Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/compute_normal.h:718:17: error: implicit instantiation of undefined template 'std::basic_ofstream<char, std::char_traits<char> >'
  std::ofstream out("computed_normals.cgal.polylines.txt");
                ^
/usr/lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/iosfwd:119:11: note: template is declared here
    class basic_ofstream;
          ^
1 error generated.
```

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/compute_normal.h [hunk 1]
```diff
@@ -182,22 +188,302 @@ compute_face_normal(typename boost::graph_traits<PolygonMesh>::face_descriptor f
...
+template <typename PolygonMesh, typename FaceNormalMap>
+void compute_face_normals(const PolygonMesh& pmesh, FaceNormalMap fnm)
+{
+  compute_face_normals(pmesh, fnm, CGAL::Polygon_mesh_processing::parameters::all_default());
+}
+
+/// \endcond
+
+template <typename GT>
+bool almost_equal(const typename GT::Vector_3 v1, const typename GT::Vector_3 v2)
```
Comment (janetournois) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/compute_normal.h]: `v1` and `v2` should be const references, isn't it?

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/compute_normal.h [hunk 2]
```diff
@@ -182,22 +188,302 @@ compute_face_normal(typename boost::graph_traits<PolygonMesh>::face_descriptor f
...
+{
+  return (CGAL::abs(1 - GT().compute_scalar_product_3_object()(v1, v2)) < 1e-10); // @tolerance
+}
+
+template <typename PolygonMesh, typename FaceNormalVector, typename K>
+bool does_enclose_other_normals(const int i, const int j, const int k,
+                                const typename K::Vector_3& nb,
+                                const typename K::FT sp_bi,
+                                std::vector<typename boost::graph_traits<PolygonMesh>::face_descriptor> incident_faces,
```
Comment (janetournois) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/compute_normal.h]: `incident_faces` should also be a const reference

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/compute_normal.h [hunk 3]
```diff
@@ -182,22 +188,302 @@ compute_face_normal(typename boost::graph_traits<PolygonMesh>::face_descriptor f
...
+    return ni;
+
+  typename GT::Vector_3 nb = traits.construct_sum_of_vectors_3_object()(ni, nj);
+
+  return nb;
+}
+
+template <typename PolygonMesh, typename FaceNormalVector, typename GT>
+std::pair<typename GT::Vector_3, bool>
+compute_most_visible_normal_2_points(std::vector<typename boost::graph_traits<PolygonMesh>::face_descriptor> incident_faces,
```
Comment (janetournois) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/compute_normal.h]: const ref

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/compute_normal.h [hunk 4]
```diff
@@ -182,22 +188,302 @@ compute_face_normal(typename boost::graph_traits<PolygonMesh>::face_descriptor f
...
+      n = nb;
+    }
+  }
+
+  return std::make_pair(n, (n != CGAL::NULL_VECTOR));
+}
+
+template <typename PolygonMesh, typename FaceNormalVector, typename GT>
+typename GT::Vector_3
+compute_most_visible_normal_3_points(std::vector<typename boost::graph_traits<PolygonMesh>::face_descriptor> incident_faces,
```
Comment (janetournois) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/compute_normal.h]: const ref

**Comment by MaelRL** [comment]
Here is some example of the change (orange is old, red is sin-based back-up, green is new):
all_large
all_zoom

**Comment by MaelRL** [comment]
Tiny time table:
```
425305 #vertices: ~1s for sin-based (red), ~1.3s for most visible (green)
1216055 #vertices: ~2.3s for sin-based (red), ~3.2 for most visible (green)
```
This seemed consistent with other meshes of similar #nv, and whether the shape was round, smooth, or with feature edges (mechanical pieces).

Defaulting to sin-based computations is very rare (<0.1%) if it happens at all (didn't happen on mechanical pieces usually)

**Comment by maxGimeno** [comment]
error in travis : /home/travis/build/CGAL/cgal/Polygon_mesh_processing/test/Polygon_mesh_processing/pmp_compute_normals_test.cpp:30:20: error: 
      no template named 'GetVertexPointMap' in namespace 'CGAL'; did you mean
      'CGAL::Polygon_mesh_processing::GetVertexPointMap'?
  typedef typename CGAL::GetVertexPointMap<Mesh>::const_type              VPMap;

**Comment by MaelRL** [comment]
It's because this PR already uses namespace changes from https://github.com/CGAL/cgal/pull/4232 so that they can be tested together.

**Comment by maxGimeno** [comment]
Ah, ok sorry.

**Comment by MaelRL** [comment]
It's my fault, I should have put a note (I did the quick fix when I merged into integration myself last week).

**Comment by maxGimeno** [comment]
I think this causes [this error](https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-16.shtml#Polygon_mesh_processing)

**Comment by MaelRL** [comment]
Looks to be some weird 32 bit platform stuff again, I tried the precision trick in 0ddece5, but it doesn't seem to have worked out.

**Comment by maxGimeno** [comment]
Your debug code is making the testsuite look like it has a bad case of jaundice :/ 
Plus, travis is unhappy with your last commit.

**Comment by MaelRL** [comment]
There are like 5 lines affected by one warning on a testsuite that doesn't matter, I'm sure it'll survive.

**Comment by maxGimeno** [comment]
More like 10, including PMP, which means it may hide other problems on other branches. 
I agree that this time of the year, it is not capital to have warnings in the testsuite, but I'd still like to know if you are currently working on that branch or if it can be removed for now, so that I can hope to flag some branches as ready to be merged.

**Comment by maxGimeno** [comment]
errors : https://cgal.geometryfactory.com/CGAL/testsuite/#Polygon_mesh_processing

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-53.shtml

**Comment by lrineau** [comment]
This PR #4065 (already merged) seems responsible for the two red results, on Linux 32 bits, in CGAL 5.0.x:  https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-I-56.shtml#Surface_mesh_approximation

I have run a `git bisect` and the results were:
```
git bisect start
# bad: [b6d5129364ac61c85c22e36326c7e236653e12d5] Merge branch 'releases/CGAL-5.0-branch'
git bisect bad b6d5129364ac61c85c22e36326c7e236653e12d5
# good: [eb9b8e6c52c2ccaace56e977f7e43a6702a6b382] Merge remote-tracking branch 'cgal/releases/CGAL-5.0-branch'
git bisect good eb9b8e6c52c2ccaace56e977f7e43a6702a6b382
# good: [610490655833d22b391d168a9bf4b1a74b69ca22] Add off file for homotopy tests
git bisect good 610490655833d22b391d168a9bf4b1a74b69ca22
# bad: [ed4cf46ee02f42fc7527332d9cdfb5deecb630bc] Merge pull request #4065 from MaelRL/PMP-New_vertex_normal_computations-GF
git bisect bad ed4cf46ee02f42fc7527332d9cdfb5deecb630bc
# good: [0567554126bb6c94b4500df9f4b4088c75f70c2d] Merge pull request #4363 from afabri/TDS_3-doc-GF
git bisect good 0567554126bb6c94b4500df9f4b4088c75f70c2d
# bad: [eeeb5929cae09c00d1b187f07a3a760c693815d2] Clean debug code
git bisect bad eeeb5929cae09c00d1b187f07a3a760c693815d2
# bad: [d9c73a1b35a9d0d65f0d7d1753cf7d32bd2a31f6] Properly test the family of normal computation functions
git bisect bad d9c73a1b35a9d0d65f0d7d1753cf7d32bd2a31f6
# good: [f05a11dde52ed54659a8b689d06e8b04e17fc9a2] Fix some internal functions not being in the 'internal' namespace
git bisect good f05a11dde52ed54659a8b689d06e8b04e17fc9a2
# bad: [b9f37cea38393796bb886ce31a9b9f613489bdeb] Hide debug code behind macros
git bisect bad b9f37cea38393796bb886ce31a9b9f613489bdeb
# skip: [bba6e57053e85fef7604638fa142990f893aedee] Misc code cleaning
git bisect skip bba6e57053e85fef7604638fa142990f893aedee
# bad: [cf357713371ca61bebc2ea650d5e65c4ca66474d] Robustify and incorporate most visible normal computations into standard fns
git bisect bad cf357713371ca61bebc2ea650d5e65c4ca66474d
# skip: [ad90fa30fd064876404dcce283684e711540b0ba] Fix potentially taking references to temporaries
git bisect skip ad90fa30fd064876404dcce283684e711540b0ba
# only skipped commits left to test
# possible first bad commit: [cf357713371ca61bebc2ea650d5e65c4ca66474d] Robustify and incorporate most visible normal computations into standard fns
# possible first bad commit: [bba6e57053e85fef7604638fa142990f893aedee] Misc code cleaning
# possible first bad commit: [ad90fa30fd064876404dcce283684e711540b0ba] Fix potentially taking references to temporaries
```

**Comment by maxGimeno** [comment]
@maelRL this is important for 4.14.3, could you have a look ?

**Comment by MaelRL** [comment]
This was merged into 5.1, how can it be important for 4.14.3 ?

**Comment by maxGimeno** [comment]
> This was merged into 5.1, how can it be important for 4.14.3 ?

Yep, I realized that just after I posted, that's why I deleted my message :S

**Pull request merged**